include/deal.II/base/numbers.h

```diff
@@ -542,6 +565,51 @@ namespace internal
   };
 }
 
+namespace numbers
+{
+
+#ifdef DEAL_II_ADOLC_WITH_ADVANCED_BRANCHING
+  // Defined in differentiation/ad/adolc_number_types.cc
+  bool
+  are_equal_valued (const adouble &value_1,
+                    const adouble &value_2);
+
+  template <typename Number>
+  bool
+  are_equal_valued (const adouble &value_1,
+                    const Number  &value_2)
+  {
+    // Use the specialised definition for two Adol-C taped types
+    return are_equal_valued(value_1, internal::NumberType<adouble>::value(value_2));
+  }
+
+  template <typename Number>
+  bool
+  are_equal_valued (const Number  &value_1,
+                    const adouble &value_2)
+  {
+    // Use the above definition
+    return are_equal_valued(value_2, value_1);
+  }
+#endif
+
+
+  template <typename Number1,typename Number2>
+  bool
+  are_equal_valued (const Number1 &value_1, const Number2 &value_2)
+  {
+    return (value_1 == internal::NumberType<Number1>::value(value_2));
```

Comment:
I see that you don't just do a `static_cast`. This may make the option I outlined above infeasible, but at least, it would be worth stating in the documentation how exactly the cast of the second argument is supposed to happen.